index.js

```diff
@@ -33,18 +33,30 @@ const main = async () => {
 
   let findings
   let status = 'IN_PROGRESS'
+  let scanInitiated = false
 
   core.debug('Checking for existing findings')
   findings = await getFindings(ECR, repository, tag)
-  if (findings === null) {
-    console.log('Requesting image scan')
-    await ECR.startImageScan({
-      imageId: {
-        imageTag: tag
-      },
-      repositoryName: repository
-    }).promise()
-    core.debug('Requested image scan')
+
+  if (
+    findings &&
+    findings.imageScanStatus &&
+    findings.imageScanStatus.status === 'IN_PROGRESS'
+  ) {
+    scanInitiated = true
+  }
+
+  if (findings === null || scanInitiated) {
```

Comment:
This logic is a little awkward.

What about this construction?
```js
if (findings === null || !scanInitiated) {
  ECR.startImageScan(...)
}

// poll until we have findings
```

Reply:
Actually, I think that logic is wrong. It might be better to pull out two boolean variables `haveResults` and `scanInitiated` (or similar). That would help clarify this high level logic.

Reply:
That would mean it would poll once even though it already could have findings?

I was using a tiny docker image at some point the scan was completed before it even reached the scan step in the workflow.

Reply:
What about the following? (Personally, I preferred how it is now, but feel free to edit however you see fit.

```js
  let findings
  let status = 'IN_PROGRESS'
  let scanInitiated = false
  let haveResults = false

  core.debug('Checking for existing findings')
  findings = await getFindings(ECR, repository, tag)

  if (
    findings &&
    findings.imageScanStatus &&
    findings.imageScanStatus.status === 'IN_PROGRESS'
  ) {
    scanInitiated = true
  }

  if (
    findings &&
    findings.imageScanFindings &&
    findings.imageScanFindings.findings &&
    !!findings.imageScanFindings.findings.length
  ) {
    haveResults = true
  }

  if (!haveResults || !scanInitiated) {
    console.log('Requesting image scan')
    await ECR.startImageScan({
      imageId: {
        imageTag: tag
      },
      repositoryName: repository
    }).promise()
    core.debug('Requested image scan')
  }

  if (!haveResults) {
    let n = 0
    while (status === 'IN_PROGRESS') {
      if (n > 0) {
        await new Promise((resolve) => {
          setTimeout(resolve, 5000)
        })
      }
      console.log('Polling ECR for image scan findings...')
      findings = await getFindings(ECR, repository, tag)
      status = findings.imageScanStatus.status
      core.debug(`Scan status: ${status}`)
      n++
    }
  }
```

Reply:
Basically the question is, how would you define `haveResults` as true? Either that there are findings in the array or that the status is completed?

Reply:
I've pushed a proposal here. What do you think? https://github.com/alexjurkiewicz/ecr-scan-image/tree/pzi-feature/poll-if-scan-already-in-progress

The goal was to make the logic more linear. I also added some handling for exceptional cases while I was in the area (unexpected status codes).

Reply:
Yeah, definitely more linear, thanks. I am currently running a workflow with your proposal, let you know how it goes.

Reply:
Great. I'll probably not be able to check back on this PR until next week, FWIW.

Reply:
Yep, that worked for our workflow.